go/general_style_guide.md

```diff
@@ -304,3 +304,123 @@ import (
 - imports are easier to read
 
 **Cons:**
+
+
+## Unit Tests
+
+Unit tests should be written when it makes sense. Writing tests first often
+helps designing new component if its shape is not enforced by framework
+requirements. Unit tests also help to verify that state mutating critical logic
+behaves correctly in both various success cases and also in error situations.
+Simple logic is not that important but sometimes it makes sense to have a simple
+test validating correct construction of bigger object so that if one forgets to
+pass configuration parameter, the validation error is caught during development
+and not in the deployment. 100% code-coverage is not something that should be
+targeted. Writing unit test also makes sense when facing a bug that could have
+been prevented by having a test for the relevant code in question.
+
+Use table-driven tests instead of separate test methods when possible. Each test
+case in table-driven test should be executed as a sub-test with
+[t.Run](https://golang.org/pkg/testing/#T.Run) as following output of `go test`
+is then easier and one can execute a single specific test case from within table
+if needed.
+
+In table-driven tests the fields of test case struct must be unexported as it's
+common practice and makes semantically more sense. When iterating test cases,
+it's common pattern to use `tc` as variable name for current element.
+
+Description field must start with `case n` where `n` is the index of the test
+starting from 0.
+
+Error matchers `func(error) bool` should be used to assert errors. This makes it
+easier to match errors returned from other packages.
+
+All test validations must be included in `t.Run(...) { }` block as technically
+each test case in table is a single test.
+
+Example function to be tested:
+```go
+
+func ClusterID(obj interface{}) (string, error) {
+	customObject, ok := obj.(SpecificType)
+	if !ok {
+		return "", microerror.Maskf(wrongTypeError, "obj must be SpecificType")
+	}
+
+	return customObject.Spec.ID, nil
+}
+```
+
+Example test:
+```go
+
+func Test_ClusterID(t *testing.T) {
+	testCases := []struct {
+		name              string
```

Comment:
Dumb question, but since I was trying this out in my own test: How do we want to order the fields in testcases?
`Name` at the top and `errorMatcher` at the bottom and alphabetical in between?

Reply:
My approach initially was just an instinct more or less based on the order of usage. `name` field appears first in all community versions I have seen so far.

My proposal would be somewhat like this:
1. `name`
2. Input parameters
3. Expected output
4. Possibly validators if needed?
5. Error matcher

Reply:
Ack, I would add setup - for example when you need to set up a service or something similar in the fake k8s api. So:
1. `name`
2. Setup config
3. Input parameters
4. Validators
5. Error matchers